### PR TITLE
feat: Pin Juju to <3.0 and update pip requirements

### PR DESCRIPTION
**Pull request by phoevos**
* Configure tox
  * Fix syntax error in `bash -c` command for update-requirements
  * Enable using the root level update-requirements to update all nested projects' requirements
  * Add the `pip-compile -U` flag
* Pin Juju to <3.0
  * Pin `python-libjuju<3.0` for integration tests to ensure compatibility with the 2.9 controller version (3.0 is not being maintained anymore)
  * Pin Juju controller to 2.9/stable for the Github runners
  * Pin Juju snap to a 2.9.42 revision (see: https://bugs.launchpad.net/juju/+bug/2024897)
* Update pip requirements for all charms

### CODE REVIEW

File: charms/kfp-api/tox.ini [hunk 1]
```diff
@@ -33,6 +33,8 @@ allowlist_externals =
     xargs
 commands = 
 ; uses 'bash -c' because piping didn't work in regular tox commands
+  pip-compile requirements.in
+  pip-compile requirements-fmt.in
```
Comment (beliaev-maksim) [inline charms/kfp-api/tox.ini:37]:
```suggestion
```

File: .github/workflows/integrate.yaml [hunk 2]
```diff
@@ -111,9 +110,12 @@ jobs:
...
           provider: microk8s
           channel: 1.24/stable
+          juju-channel: 2.9/stable
           charmcraft-channel: latest/candidate
-          # Pinning juju agent to 2.9.42 to keep compatibility with pythonlib-juju<3
-          bootstrap-options: --agent-version="2.9.42"
+
+      - name: Refresh Juju snap
+        run: |
+          sudo snap refresh juju --revision 22345
```
Comment (beliaev-maksim) [inline .github/workflows/integrate.yaml:118]:
```suggestion
      - name: Refresh Juju snap
        run: |
          # remove when fixed https://bugs.launchpad.net/juju/+bug/2024897
          sudo snap refresh juju --revision 22345
```

File: charms/kfp-persistence/src/charm.py [hunk 3]
```diff
@@ -122,6 +122,8 @@ def _get_interfaces(self):
             raise CheckFailedError(str(err), WaitingStatus)
         except NoCompatibleVersions as err:
             raise CheckFailedError(str(err), BlockedStatus)
+        except RelationDataError as err:
+            raise CheckFailedError(str(err), BlockedStatus)
```
Comment (beliaev-maksim) [inline charms/kfp-persistence/src/charm.py:126]: are you sure it is Blocked?

what can user do to unblock it?

File: charms/kfp-persistence/tox.ini [hunk 4]
```diff
@@ -33,7 +33,9 @@ allowlist_externals =
     xargs
 commands = 
 ; uses 'bash -c' because piping didn't work in regular tox commands
-  bash -c \'find . -type f -name "requirements*.in" | xargs --replace=\{\} pip-compile --resolver=backtracking \{\}'
+  pip-compile requirements.in
+  pip-compile requirements-fmt.in
```
Comment (beliaev-maksim) [inline charms/kfp-persistence/tox.ini:37]:
```suggestion
```

File: charms/kfp-profile-controller/tox.ini [hunk 5]
```diff
@@ -33,7 +33,9 @@ allowlist_externals =
     xargs
 commands = 
 ; uses 'bash -c' because piping didn't work in regular tox commands
-  bash -c \'find . -type f -name "requirements*.in" | xargs --replace=\{\} pip-compile --resolver=backtracking \{\}'
+  pip-compile requirements.in
+  pip-compile requirements-fmt.in
```
Comment (beliaev-maksim) [inline charms/kfp-profile-controller/tox.ini:37]:
```suggestion
```

File: charms/kfp-schedwf/tox.ini [hunk 6]
```diff
@@ -33,7 +33,9 @@ allowlist_externals =
     xargs
 commands = 
 ; uses 'bash -c' because piping didn't work in regular tox commands
-  bash -c \'find . -type f -name "requirements*.in" | xargs --replace=\{\} pip-compile --resolver=backtracking \{\}'
+  pip-compile requirements.in
+  pip-compile requirements-fmt.in
```
Comment (beliaev-maksim) [inline charms/kfp-schedwf/tox.ini:37]:
```suggestion
```

File: charms/kfp-ui/src/charm.py [hunk 7]
```diff
@@ -256,6 +256,8 @@ def _get_interfaces(self):
             raise CheckFailedError(str(err), WaitingStatus)
         except NoCompatibleVersions as err:
             raise CheckFailedError(str(err), BlockedStatus)
+        except RelationDataError as err:
+            raise CheckFailedError(str(err), BlockedStatus)
```
Comment (beliaev-maksim) [inline charms/kfp-ui/src/charm.py:260]: same here

File: charms/kfp-ui/tox.ini [hunk 8]
```diff
@@ -33,7 +33,9 @@ allowlist_externals =
     xargs
 commands = 
 ; uses 'bash -c' because piping didn't work in regular tox commands
-  bash -c \'find . -type f -name "requirements*.in" | xargs --replace=\{\} pip-compile --resolver=backtracking \{\}'
+  pip-compile requirements.in
+  pip-compile requirements-fmt.in
```
Comment (beliaev-maksim) [inline charms/kfp-ui/tox.ini:37]:
```suggestion
```

File: charms/kfp-viewer/tox.ini [hunk 9]
```diff
@@ -33,7 +33,9 @@ allowlist_externals =
     xargs
 commands = 
 ; uses 'bash -c' because piping didn't work in regular tox commands
-  bash -c \'find . -type f -name "requirements*.in" | xargs --replace=\{\} pip-compile --resolver=backtracking \{\}'
+  pip-compile requirements.in
+  pip-compile requirements-fmt.in
```
Comment (beliaev-maksim) [inline charms/kfp-viewer/tox.ini:37]:
```suggestion
```

File: charms/kfp-viz/src/charm.py [hunk 10]
```diff
@@ -126,6 +131,8 @@ def _get_interfaces(self):
             raise CheckFailedError(str(err), WaitingStatus)
         except NoCompatibleVersions as err:
             raise CheckFailedError(str(err), BlockedStatus)
+        except RelationDataError as err:
+            raise CheckFailedError(str(err), BlockedStatus)
```
Comment (beliaev-maksim) [inline charms/kfp-viz/src/charm.py:135]: same

File: charms/kfp-viz/tox.ini [hunk 11]
```diff
@@ -33,7 +33,9 @@ allowlist_externals =
     xargs
 commands = 
 ; uses 'bash -c' because piping didn't work in regular tox commands
-  bash -c \'find . -type f -name "requirements*.in" | xargs --replace=\{\} pip-compile --resolver=backtracking \{\}'
+  pip-compile requirements.in
+  pip-compile requirements-fmt.in
```
Comment (beliaev-maksim) [inline charms/kfp-viz/tox.ini:37]:
```suggestion
```

**Review by beliaev-maksim — CHANGES_REQUESTED** [review]

File: charms/kfp-persistence/src/charm.py (hunk 3 above)
Comment (phoevos) [reply charms/kfp-persistence/src/charm.py:126]: I'm not, it's just what we do in the KFP API charm: https://github.com/canonical/kfp-operators/blob/b19cf459b6974e839280f52501b3268ad37647d7/charms/kfp-api/src/charm.py#L338-L339

File: charms/kfp-api/tox.ini (hunk 1 above)
Comment (phoevos) [reply charms/kfp-api/tox.ini:37]: We do this in other places too, e.g:
* https://github.com/canonical/notebook-operators/blob/a8f0aefcc16e9242742c329c08ea16bfe4b8733c/charms/jupyter-controller/tox.ini#L36-L38
* https://github.com/canonical/oidc-gatekeeper-operator/blob/8f98b252c69d32aa3b7fd1b8ce9fbd042d49cd79/tox.ini#L36-L38

File: charms/kfp-persistence/src/charm.py (hunk 3 above)
Comment (phoevos) [reply charms/kfp-persistence/src/charm.py:126]: > related?

I don't know, there's nothing on that issue regarding what exactly happened and how you ended up there 😛

Comment (phoevos) [reply charms/kfp-persistence/src/charm.py:126]: What would you expect the status to be there?

Comment (beliaev-maksim) [reply charms/kfp-persistence/src/charm.py:126]: we cannot just set waiting without issue handling, since what will we wait :)

Comment (phoevos) [reply charms/kfp-persistence/src/charm.py:126]: Will add some context in the linked issue. Let's keep this as is for now, until we have a solid solution.

**Review by beliaev-maksim — APPROVED** [review]
LGTM if CI passes

**Pull request merged**